Resolves #139

kameshsampath opened the pull request:
### What does this PR do?
When `MAVEN_MIRROR_URL` is set in the container as environment variable, then the value of the environment variable will be set as mirror in the `$HOME/.m2/settings.xml`. This enables for faster builds and will allow che to be run in restricted environments
### What issues does this PR fix or reference?
Issue #139 

### Previous behavior
Each build will pull the artifacts and dependencies, this will be an issue when che is running in restricted environment when che can't reach internet

### New behavior
Support for  `MAVEN_MIRROR_URL`as environment variable, which can act as maven mirror url in  in the `$HOME/.m2/settings.xml`.

Please review [Che's Contributing Guide](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md) for best practices.

### Tests written?
No
### Docs updated?
N.A

benoitf commented on recipes/centos_spring_boot/Dockerfile:
```diff
@@ -16,7 +16,12 @@ ARG JUNIT_VERSION=4.12
-    sudo rm -f /tmp/install_spring_boot_dependencies.sh
+    sudo rm -f /tmp/install_spring_boot_dependencies.sh && \
+    sudo chown user:user ${HOME}/run.sh && \
+    sudo chmod +x ${HOME}/run.sh
```
we could merge the two chown and chmod instructions as we can do `chmod <options> file1 file2`

kameshsampath replied:
sorry @benoitf - i never used that command :(

benoitf commented on recipes/centos_spring_boot/Dockerfile:
```diff
@@ -16,7 +16,12 @@ ARG JUNIT_VERSION=4.12
+    sudo chown user:user ${HOME}/run.sh && \
+    sudo chmod +x ${HOME}/run.sh
+
+CMD ["sh","-c","${HOME}/run.sh"]
```
AFAIK the run instructions need to have infinite loop
Maybe you want entrypoint instead of command ?

also there is a missing newline

kameshsampath replied:
I would always avoid ENTRYPOINT as that takes the user to customize anything later by passing parameters - let me see if I can add the infinite looping there

benoitf commented on recipes/centos_spring_boot/run.sh:
```diff
@@ -0,0 +1,23 @@
+set -e
+
+# insert settings for mirrors/repository managers into settings.xml if supplied
+function configure_mirrors {
```
only `configure_mirrors() {` ?

kameshsampath replied:
yes thats it for now  - we might need a PROXY config as well. but am not adding that to this as I can have separate/issue/pr for it

benoitf replied:
I only mean : replace
```
 function abcd { 
```
by
```
abcd() {
```

kameshsampath replied:
yeah got it done after using the shellcheck tool . thanks for clarification

benoitf commented on recipes/centos_spring_boot/run.sh:
```diff
@@ -0,0 +1,23 @@
+}
+
+# check for the MAVEN_MIRROR_URL env variable, if its available then set maven mirrors in $HOME/.m2/settings.xml
+configure_mirrors
```
missing new line

Also please check on https://www.shellcheck.net/

kameshsampath replied:
Thanks for the site, has been using yamllint/jsonlint for longtime never knew we have tool like this very handy

benoitf commented on recipes/centos_vertx/Dockerfile:
```diff
@@ -16,7 +16,12 @@ ARG VERTX_VERSION=3.4.2
+    sudo chown user:user ${HOME}/run.sh && \
+    sudo chmod +x ${HOME}/run.sh
+
+CMD ["sh","-c","${HOME}/run.sh"]
```
newline + same comments than before

kameshsampath replied:
here you want me to avoid new line or add new line after Line 27 ?

benoitf commented on recipes/centos_vertx/run.sh:
```diff
@@ -0,0 +1,23 @@
+}
+
+# check for the MAVEN_MIRROR_URL env variable, if its available then set maven mirrors in $HOME/.m2/settings.xml
+configure_mirrors
```
newline + same comments than before

benoitf commented on recipes/centos_wildfly_swarm/run.sh:
```diff
@@ -0,0 +1,23 @@
+}
+
+# check for the MAVEN_MIRROR_URL env variable, if its available then set maven mirrors in $HOME/.m2/settings.xml
+configure_mirrors
```
newline + same comments than before

kameshsampath commented on recipes/centos_wildfly_swarm/run.sh:
```diff
@@ -0,0 +1,25 @@
+# check for the MAVEN_MIRROR_URL env variable, if its available then set maven mirrors in $HOME/.m2/settings.xml
+configure_mirrors
+
+while :; do sleep 1; done
```
@benoitf  in all the `run.sh` i have added this infinite loop to keep the docker live. As told in the comment I would avoid `ENTRTPOINT` as then user can't do `docker exec -it mycontainer <some-command>`.

benoitf replied:
maybe use the same command than
https://github.com/eclipse/che-dockerfiles/blob/master/recipes/stack-base/centos/Dockerfile#L62 ?
```
tail -f /dev/null
```

benoitf commented on recipes/centos_spring_boot/Dockerfile:
```diff
@@ -16,7 +16,10 @@ ARG JUNIT_VERSION=4.12
 COPY install_spring_boot_dependencies.sh /tmp/
-RUN sudo chown user:user /tmp/install_spring_boot_dependencies.sh && \
-    sudo chmod a+x /tmp/install_spring_boot_dependencies.sh && \
+COPY ./run.sh ${HOME}/
```
question: do we really need to leave the command in ${HOME} folder ?
user may want to run that script as well in the terminal ?

kameshsampath commented:
@benoitf am not specific about it, a possible scenario say if the developer on the Che WS gets to terminal and want to update the MAVEN_MIRROR_URL for some testing - a very usual scenario in the enterprises, then if we have the script in his home folder he can run it any time we wants.  but as a convention or process you want to move to temp I don't mind.

benoitf commented:
@kameshsampath discussion is open so others may have different views. I approved as it's not related directly to what the PR is doing

benoitf commented:
@kameshsampath did you have tested these images on OpenShift or OSIO ?

kameshsampath commented:
@benoitf i have done testing of it locally with DOCKER connected to minishift and normal docker for mac

kameshsampath commented:
@l0rd I thought about it, but then the entrypoint.sh comes from base centos image which is used by other images where maven is irrelavant, so was reluctant in modifying the entrypoint.sh so felt duplicated the code.

l0rd commented:
Maven is relevant in image centos_jdk8 and in all images based on it.

kameshsampath commented:
@l0rd  - my bad it meant the base centos image form where centos-jdk8 is based where entrypoint.sh is defined.

I see option here, overiding `entrypoint.sh` from base centos image in centos-jdk8 image and then add the run logic there and remove it from other base images .

thoughts ?

kameshsampath commented:
@l0rd - i have got another idea as well, what if we we override entrypoint.sh via centos-jdk8 image? if you feel thats good then I can have the PR patched for it and remove redundant `run.sh`

kameshsampath commented:
@l0rd @benoitf  please check the latest commit, i have avoided the redundancy of maven configuration and moved it to base image `centos-jdk8`

I have also built a small testing for all these images not added as part of the PR but you can see them here https://github.com/kameshsampath/docker-image-testing-demo , where i have added two test cases one without env variable and one with env variable

PS: those test cases does not cover all the usecases but gives an idea of how we can test all our docker images, if you guys find it interesting, we can have a issue and PR for it separately.

benoitf commented on recipes/centos_spring_boot/Dockerfile:
```diff
@@ -16,6 +16,7 @@ ARG JUNIT_VERSION=4.12
 ENV SPRING_BOOT_GROUP=org.springframework.boot
 
 COPY install_spring_boot_dependencies.sh /tmp/
+
```
you can remove this additional line

kameshsampath replied:
done!

benoitf commented on recipes/centos_vertx/Dockerfile:
```diff
@@ -16,6 +16,7 @@ ARG VERTX_VERSION=3.4.2
 ENV VERTX_GROUPID=io.vertx 
 
 COPY install-vertx-dependencies.sh /tmp/
+
```
you can remove this additional line

kameshsampath replied:
done!

benoitf commented on recipes/centos_wildfly_swarm/Dockerfile:
```diff
@@ -14,6 +14,7 @@ LABEL che:server:8080:ref=wildfly che:server:8080:protocol=http
 ARG SWARM_VERSION=2017.10.0
 
 COPY install-swarm-dependencies.sh /tmp/
+
```
you can remove this additional line

kameshsampath replied:
done!

benoitf commented on recipes/centos_jdk8/contrib/run.sh:
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
```
is it bash or it only requires sh ?

kameshsampath replied:
sh shud be ok

benoitf commented:
@riuvshin are you interested in these QA tests ?

eivantsov commented:
@benoitf we have no qa tests that use this particular image

benoitf commented:
@eivantsov yes but that would be nice to have QA on the docker images, so we can automatically test the PRs

eivantsov commented:
@benoitf tests like what? If the image runs? Some particular functionality?

kameshsampath commented:
@eivantsov - please check this demo https://github.com/kameshsampath/docker-image-testing-demo 
in that i was testing centos-jdk8 images

* like right JDK version installed
* right maven version installed 
* ensure that settings.xml added to right directory 
* another test case within that tests to see if the image is updated with right maven mirror url  when we pass the env var

benoitf commented:
@eivantsov yes like testing if it is working, if expected content is there, if some custom ENV variables are working as expected, etc

l0rd commented on recipes/centos_jdk8/contrib/run.sh:
```diff
@@ -0,0 +1,29 @@
+
+# insert settings for mirrors/repository managers into settings.xml if supplied
+configure_mirrors() {
+  echo "Setting maven mirror from $MAVEN_MIRROR_URL"
```
`MAVEN_MIRROR_URL` may be not set here and you would log:

```
Setting maven mirror from
```
You may want to move this log inside the `if` clause body.

kameshsampath replied:
oops!!! sorry @l0rd  - i was having that for test missed to removed it before i pushed

kameshsampath commented:
@l0rd  - let me know on it, will kind of send a fresh PR . thanks @l0rd ,@benoitf , @eivantsov for reviewing and merging

l0rd commented:
@benoitf @kameshsampath @eivantsov @riuvshin +1 for automate testing of images

eivantsov commented:
@l0rd sure testing images is good, but will that mean that each image has its individual test?

l0rd commented:
@kameshsampath the adjustment was the log you have just fixed it 👍 

@eivantsov I would start with one image, include the tests in the CI and see how it goes. If we find it useful we should write one test file for each image that appears in `stacks.json`.

If we are not going to tests the images users will do it 😄

kameshsampath commented:
@eivantsov if you check my sample repo i already wrote a test for centos-jdk8,  we can reuse it. I

> @eivantsov I would start with one image, include the tests in the CI and see how it goes. If we find it useful we should write one test file for each image that appears in stacks.json.
yes, right now I have added it to centos-jdk? shall we have a ticket to discuss on this and start flushing it out if its worth it .. I hope so ;)
